### PR TITLE
Unit Tests for Roles, Invites, Subscriptions-related Routes. 

### PR DESCRIPTION
**Pull request by zakhap**
## Description
These are a bunch of unit tests and some route fixes discovered during testing. Pretty much covers ever route I've worked with since January with a few outstanding (with tests) in another PR. 

## Motivation and Context
Part of the ongoing saga of testing our code!

## How has this been tested?
`yarn test-api` > coverage seems to be +90% for all of these new tests.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [x] yes, and they are tested: +90%

### CODE REVIEW

**Comment by drewstone** [comment]
Seems the merge conflicts re-introduced some issues to more tests. These should be fixed before merging this in.

**Comment by zakhap** [comment]
> Seems the merge conflicts re-introduced some issues to more tests. These should be fixed before merging this in.

I fixed these bugs! It was an issue with creating threads with more than one tag that I think was structurally introduced with the reduction of thread tags to 1, but it was a simple quick fix.

**Comment by drewstone** [comment]
Awesome job @zakhap !

File: test/unit/api/threads.spec.ts [hunk 1]
```diff
@@ -238,6 +240,50 @@ describe('Thread Tests', () => {
...
+        .set('Accept', 'application/json')
+        .query({
+          chain,
+          jwt: adminJWT,
+        });
+      expect(res.body.result).to.not.be.null;
+      expect(res.body).to.not.be.null;
+      expect(res.body.status).to.be.equal('Success');
+    });
+    it('should pass with public community', async () => {
```
Comment (drewstone) [inline test/unit/api/threads.spec.ts:256]: I would change these descriptions to be more specific, saying the test should pass is not clear for a reader what this is testing.

File: test/unit/api/threads.spec.ts [hunk 2]
```diff
@@ -238,6 +240,50 @@ describe('Thread Tests', () => {
     });
   });
 
+  describe('/bulkThreads', () => {
+    it('should pass with public chain', async () => {
```
Comment (drewstone) [inline test/unit/api/threads.spec.ts:244]: same here wit description, more clarity on what this is testing is the move.

File: test/unit/api/subscriptions.spec.ts [hunk 3]
```diff
@@ -255,18 +255,39 @@ describe('Subscriptions Tests', () => {
...
+          .post('/api/viewNotifications')
+          .set('Accept', 'application/json')
+          .send({ jwt: jwtToken, unread_only: true, });
+        expect(res.body).to.not.be.null;
+        expect(res.body.status).to.be.equal('Success');
+        expect(res.body.result.length).to.be.greaterThan(0);
+        notifications = res.body.result;
+      });
+
+      it('should pass with active_only property', async () => {
```
Comment (drewstone) [inline test/unit/api/subscriptions.spec.ts:281]: here too

File: test/unit/api/subscriptions.spec.ts [hunk 4]
```diff
@@ -255,18 +255,39 @@ describe('Subscriptions Tests', () => {
...
+          .post('/api/viewNotifications')
+          .set('Accept', 'application/json')
+          .send({ jwt: jwtToken });
+        expect(res.body).to.not.be.null;
+        expect(res.body.status).to.be.equal('Success');
+        expect(res.body.result.length).to.be.greaterThan(0);
+        notifications = res.body.result;
+      });
+
+      it('should pass with unread_only property', async () => {
```
Comment (drewstone) [inline test/unit/api/subscriptions.spec.ts:270]: here too

File: test/unit/api/subscriptions.spec.ts [hunk 5]
```diff
@@ -255,18 +255,39 @@ describe('Subscriptions Tests', () => {
...
-      expect(thread).to.not.be.null;
-      const res = await chai.request(app)
-        .post('/api/viewNotifications')
-        .set('Accept', 'application/json')
-        .send({ jwt: jwtToken });
-      expect(res.body).to.not.be.null;
-      expect(res.body.status).to.be.equal('Success');
-      expect(res.body.result.length).to.be.greaterThan(0);
-      notifications = res.body.result;
+    describe('/viewNotifications route tests', () => {
```
Comment (drewstone) [inline test/unit/api/subscriptions.spec.ts:258]: here too

File: test/unit/api/roles.spec.ts [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        .set('Accept', 'application/json')
+        .query({
+          community,
+          jwt: jwtToken
+        });
+      expect(res.body.status).to.be.equal('Success');
+      expect(res.body.result.length).to.be.greaterThan(0);
+    });
+
+    it('should fail if community not visible to user', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:303]: "should fail to grab bulk members if community is not visible to user"

File: test/unit/api/roles.spec.ts [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+          chain,
+          address_id: memberAddressId,
+        });
+      expect(res.body.error).to.not.be.null;
+      expect(res.body.error).to.be.equal(deleteErrors.RoleDNE);
+    });
+  });
+
+  describe('/bulkMembers route test', () => {
+    it('should pass with standard input', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:291]: "should grab bulk members for a public community" or "should successfully grab bulk members for a public community"

File: test/unit/api/roles.spec.ts [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+        .set('Accept', 'application/json')
+        .send({
+          jwt: memberJwt,
+          community,
+        });
+      expect(res.body.error).to.not.be.null;
+      expect(res.body.error).to.be.equal(deleteErrors.InvalidAddress);
+    });
+
+    it('should fail with invalid address_id', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:263]: "should fail to delete role with an invalid address_id"

File: test/unit/api/roles.spec.ts [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+        .set('Accept', 'application/json')
+        .send({
+          jwt: memberJwt,
+          community,
+          address_id: memberAddressId,
+        });
+      expect(res.body.status).to.be.equal('Success');
+    });
+
+    it('should fail without address_id', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:251]: more descriptive

File: test/unit/api/roles.spec.ts [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+          jwt: jwtToken,
+          community,
+          address: newUserAddress,
+          new_role: 'commander in chief',
+        });
+      expect(res.body.error).to.not.be.null;
+      expect(res.body.error).to.be.equal(upgradeErrors.InvalidRole);
+    });
+
+    it('should fail when not an admin requests', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:175]: "should fail when a non-admin upgrades a member"

File: test/unit/api/roles.spec.ts [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+        .send({
+          jwt: jwtToken,
+          community,
+          address_id: loggedInAddrId,
+        });
+      expect(res.body.error).to.not.be.null;
+      expect(res.body.error).to.be.equal(createErrors.RoleAlreadyExists);
+    });
+
+    it('should fail without address_id', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:72]: "should fail to create role without address_id"

File: test/unit/api/roles.spec.ts [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+          jwt: jwt.sign({ id: user.user_id, email: user.email }, JWT_SECRET),
+          community,
+          address_id: user.address_id,
+        });
+      expect(res.body.status).to.be.equal('Success');
+      expect(res.body.result.address_id).to.be.equal(user.address_id);
+      expect(res.body.result.offchain_community_id).to.be.equal(community);
+    });
+
+    it('should fail on joining public community a second time', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:59]: "should fail to create duplicate role for a public community a user is a member of"

File: test/unit/api/roles.spec.ts [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+    adminUserId = res.user_id;
+    const isAdmin = await modelUtils.assignRole({
+      address_id: res.address_id,
+      chainOrCommObj: { offchain_community_id: community },
+      role: 'admin',
+    });
+  });
+
+  describe('/createRole route tests', () => {
+    it('should pass on joining public community', async () => {
```
Comment (drewstone) [inline test/unit/api/roles.spec.ts:44]: more descriptive, please no "should pass" descriptors

File: test/unit/api/threads.spec.ts (hunk 1 above)
Comment (zakhap) [reply test/unit/api/threads.spec.ts:256]: "should return bulk threads for a public chain"

File: test/unit/api/threads.spec.ts (hunk 2 above)
Comment (zakhap) [reply test/unit/api/threads.spec.ts:244]: "should return bulk threads for a public chain"

File: test/unit/api/subscriptions.spec.ts (hunk 5 above)
Comment (zakhap) [reply test/unit/api/subscriptions.spec.ts:258]: "/viewNotifications: return notifications to user"

**Comment by zakhap** [comment]
@drewstone went through the descriptions and updated them to be more descriptive!

**Pull request merged**